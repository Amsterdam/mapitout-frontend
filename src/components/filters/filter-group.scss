@import "../../style/variables";

.filter-group {
  width: 100%;
}

.toggle-expand {
  display: flex;
  flex-direction: row;
  align-items: center;
  height: 48px;
  width: 100%;
  padding: 0 24px;
  border: 0 none;
  outline: 0 none;
  background-color: $greyscale-2;
  cursor: pointer;
  font-size: 14px;

  .icon {
    width: 28px;
    height: 28px;
    margin-right: 16px;
  }

  .icon-toggle {
    width: 8px;
    height: 4px;
    margin-left: 16px;
    transform: rotate(180deg);

    .collapsed & {
      transform: rotate(0);
    }
  }

  .name {
    flex-grow: 1;
    text-align: left;
  }
}

.group {
  list-style: none;
  padding: 0;
  margin: 0;

  .item {
    border-bottom: 2px solid $greyscale-2;
  }
}

.filter::v-deep {
  background-color: white;

  .icon {
    visibility: hidden;
  }
}
