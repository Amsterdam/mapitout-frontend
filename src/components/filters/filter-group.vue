<template>
  <div class="filter-group" :class="{ collapsed: !expanded }">
    <button class="toggle-expand" @click="onClickExpand">
      <component :is="group.iconComponent" class="icon" />
      <span class="name">{{ group.name }}</span>
      <caret-down class="icon-toggle" />
    </button>
    <ul class="group" v-if="expanded">
      <li class="item" v-for="filter in group.children" :key="filter.value">
        <component-filter :filter="filter" />
      </li>
    </ul>
  </div>
</template>
<style scoped lang="scss" src="./filter-group.scss"></style>
<script>
import IconBus from "@/assets/icons/IconBus.svg?inline";
import IconTrain from "@/assets/icons/IconTrain.svg?inline";
import IconTram from "@/assets/icons/IconTram.svg?inline";
import IconSubway from "@/assets/icons/IconSubway.svg?inline";
import IconEducation from "@/assets/icons/IconEducation.svg?inline";
import CaretDown from "@/assets/CaretDown.svg?inline";

import ComponentFilter from "./filter.vue";

export default {
  components: {
    IconBus,
    IconTrain,
    IconTram,
    IconSubway,
    IconEducation,
    CaretDown,
    ComponentFilter
  },
  props: {
    group: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      expanded: false
    };
  },

  methods: {
    onClickExpand() {
      this.expanded = !this.expanded;
    }
  }
};
</script>
