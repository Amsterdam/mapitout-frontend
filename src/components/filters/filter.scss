@import "../../style/variables";

.filter {
  height: 48px;
  display: flex;
  align-items: center;
  background-color: $greyscale-2;
  padding: 0 24px;
}

.icon {
  width: 28px;
  height: 28px;
  margin-right: 16px;
}

.name {
  flex-grow: 1;
  font-size: 14px;
}

button.name {
  border: 0 none;
  outline: 0 none;
  background-color: transparent;
  text-decoration: underline;
  text-align: left;
  padding: 0;
  margin: 0;
  cursor: pointer;
}

.toggle {
  position: relative;
  height: 28px;
  width: 50px;
  border: 0 none;
  background-color: rgba($greyscale-1, 0.2);
  outline: none;
  cursor: pointer;
  border-radius: 14px;
  transition: background-color 0.1s ease-in-out;

  &::after {
    content: " ";
    display: block;
    background-color: white;
    width: 22px;
    height: 22px;
    border-radius: 11px;
    position: absolute;
    left: 3px;
    top: 3px;
    transition: transform 0.1s ease-in-out;
  }

  &.selected {
    background-color: $greyscale-1;

    &::after {
      transform: translateX(22px);
    }
  }
}
