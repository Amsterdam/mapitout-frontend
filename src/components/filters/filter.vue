<template>
  <div class="filter">
    <component :is="filter.iconComponent" class="icon" />
    <button class="name" v-if="hasChildren(filter.id)" @click="onClickName">
      {{ filter.name }}
    </button>
    <span class="name" v-else>{{ filter.name }}</span>
    <button
      class="toggle"
      :class="{ selected: filter.selected }"
      @click="onClickToggle"
    ></button>
  </div>
</template>
<style lang="scss" scoped src="./filter.scss"></style>
<script>
import { mapActions, mapGetters, mapMutations } from "vuex";

import IconBus from "@/assets/icons/IconBus.svg?inline";
import IconTrain from "@/assets/icons/IconTrain.svg?inline";
import IconTram from "@/assets/icons/IconTram.svg?inline";
import IconSubway from "@/assets/icons/IconSubway.svg?inline";
import IconEducation from "@/assets/icons/IconEducation.svg?inline";

export default {
  components: {
    IconBus,
    IconTrain,
    IconTram,
    IconSubway,
    IconEducation
  },

  props: {
    filter: {
      type: Object,
      required: true
    }
  },

  computed: {
    ...mapGetters("filters", ["hasChildren"])
  },

  methods: {
    ...mapActions("filters", {
      toggleFilter: "toggle"
    }),
    ...mapMutations("filters", ["setRootFilter"]),

    onClickToggle() {
      this.toggleFilter(this.filter);
    },

    onClickName() {
      this.setRootFilter(this.filter);
    }
  }
};
</script>
