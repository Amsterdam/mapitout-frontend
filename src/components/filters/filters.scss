@import "../../style/variables";

.panel-filters {
  @media (min-width: $breakpoint-tablet-portrait) {
    box-shadow: 0 2px 8px 0 lighten($greyscale-1, 50);
  }
}

.back {
  outline: 0 none;
  border: 0 none;
  padding: 0;
  background-color: transparent;
  cursor: pointer;
}

.header {
  min-height: 50px;
  background: $greyscale-0;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 0 48px;

  .title {
    text-transform: uppercase;
    color: $greyscale-2;
    font-size: 12px;
    font-weight: 500;
    display: flex;
    align-items: center;
    line-height: 1;
    flex-grow: 1;

    .icon {
      width: 16px;
      height: 16px;
      margin-right: 12px;
      color: $greyscale-2;

      &.icon-filters {
        @media (max-width: $breakpoint-tablet-portrait) {
          position: absolute;
          right: 48px;
        }
      }
    }
  }
}

.body {
  background: white;
  max-height: 100%;
  overflow-y: auto;
}

.items {
  padding: 20px;
  margin: 0;
  list-style: none;

  .item {
    margin: 10px 0;

    &:first-child {
      margin-top: 0;
    }

    &:last-child {
      margin-bottom: 0;
    }
  }
}
