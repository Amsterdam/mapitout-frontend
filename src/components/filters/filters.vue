<template>
  <div class="panel-filters">
    <div class="header">
      <div class="title">
        <button class="back" v-if="rootFilter !== null" @click="onClickBack">
          <icon-arrow-left class="icon" />
        </button>
        <icon-filters class="icon icon-filters" />
        <span>{{ rootFilter === null ? "Filters" : rootFilter.name }}</span>
      </div>
    </div>
    <div class="body">
      <ul class="items">
        <li class="item" v-for="item in filtersForRoot" :key="item.name">
          <component-filter-group v-if="item.isGroup" :group="item" />
          <component-filter v-if="item.isFilter" :filter="item" />
        </li>
      </ul>
    </div>
  </div>
</template>
<style lang="scss" scoped src="./filters.scss"></style>
<script>
import { mapActions, mapState, mapMutations } from "vuex";

import IconFilters from "@/assets/icons/IconFilters.svg?inline";
import IconArrowLeft from "@/assets/icons/IconArrowLeft.svg?inline";

import ComponentFilter from "./filter.vue";
import ComponentFilterGroup from "./filter-group.vue";

export default {
  components: {
    ComponentFilter,
    ComponentFilterGroup,
    IconFilters,
    IconArrowLeft
  },

  props: {
    value: Boolean
  },

  computed: {
    ...mapState("filters", {
      rawFilters: "filters",
      rootFilter: "rootFilter"
    }),

    filtersForRoot: function() {
      const parentId = this.rootFilter === null ? 0 : this.rootFilter.id;

      const childFilters = this.rawFilters.filter(
        filter => filter.parentId === parentId
      );

      const filters = childFilters
        .filter(filter => typeof filter.group === "undefined")
        .map(filter => ({ ...filter, isFilter: true }));

      const groups = childFilters
        .filter(filter => typeof filter.group !== "undefined")
        .reduce((acc, filter) => {
          if (typeof acc[filter.group] === "undefined") {
            acc[filter.group] = {
              name: filter.group,
              iconComponent: filter.iconComponent,
              isGroup: true,
              children: [filter]
            };
          } else {
            acc[filter.group].children.push(filter);
          }
          return acc;
        }, {});

      return filters
        .concat(Object.values(groups))
        .sort((a, b) => (a.name > b.name ? 1 : -1));
    }
  },

  methods: {
    ...mapActions("filters", {
      updateFilters: "update"
    }),

    ...mapMutations("filters", ["setRootFilter"]),

    onClickBack() {
      this.setRootFilter(null);
    }
  }
};
</script>
