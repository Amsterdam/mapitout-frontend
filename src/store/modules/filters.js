export const mutations = {
  toggle(state, { id, selected }) {
    state.filters = state.filters.map(filter => {
      if (filter.id !== id && filter.parent !== id) {
        return filter;
      }

      return { ...filter, selected };
    });
  },

  update(state, filters) {
    state.filters = filters;
  }
};

export const actions = {
  toggle({ commit }, payload) {
    commit("toggle", payload);
  }
};

export default {
  namespaced: true,
  state: {
    filters: [
      {
        id: 0,
        name: "Stations",
        value: "Station",
        category: null,
        root: true,
        selected: false,
        parent: null,
        children: null,
        icon: "icon-bus"
      },
      // {
      //   id: 9,
      //   name: "Train Station",
      //   value: "Train Station",
      //   category: null,
      //   root: false,
      //   selected: false,
      //   parent: 0,
      //   children: null,
      //   icon: "icon-train"
      // },
      // {
      //   id: 2,
      //   name: "Bus Station",
      //   value: "Bus Station",
      //   category: null,
      //   root: false,
      //   selected: false,
      //   parent: 0,
      //   children: null,
      //   icon: "icon-bus"
      // },
      // {
      //   id: 1,
      //   name: "Tram Station",
      //   value: "Tram Station",
      //   category: null,
      //   root: false,
      //   selected: false,
      //   parent: 0,
      //   children: null,
      //   icon: "icon-tram"
      // },
      // {
      //   id: 10,
      //   name: "Subway Station",
      //   value: "Subway Station",
      //   category: null,
      //   root: false,
      //   selected: false,
      //   parent: 0,
      //   children: null,
      //   icon: "icon-subway"
      // },
      {
        id: 3,
        name: "Schools",
        value: "School",
        category: null,
        root: true,
        selected: false,
        parent: null,
        children: [4, 5, 6, 7, 8],
        icon: "icon-education"
      },
      {
        id: 4,
        name: "Age range 4-11",
        value: "Age range 4-11",
        category: null,
        root: false,
        selected: false,
        parent: 3,
        children: null,
        icon: "icon-education"
      },
      {
        id: 5,
        name: "Age range 4-18",
        value: "Age range 4-18",
        category: null,
        root: false,
        selected: false,
        parent: 3,
        children: null,
        icon: "icon-education"
      },
      {
        id: 6,
        name: "Primary education",
        value: "Primary education",
        category: null,
        root: false,
        selected: false,
        parent: 3,
        children: null,
        icon: "icon-education"
      },
      {
        id: 7,
        name: "Secondary education",
        value: "Secondary education",
        category: null,
        root: false,
        selected: false,
        parent: 3,
        children: null,
        icon: "icon-education"
      }
    ]
  },
  mutations,
  actions
};
