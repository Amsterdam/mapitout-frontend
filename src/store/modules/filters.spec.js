import { getters, mutations, actions } from "./filters";

import fixtureFilters from "../../__fixtures__/filters";
import fixtureFiltersWithSelectedTypes from "../../__fixtures__/filtersWithSelectedTypes";
import fixtureFiltersWithSelectedSubTypes from "../../__fixtures__/filtersWithSelectedSubTypes";

describe("filters store module", () => {
  describe("getters", () => {
    describe("selectedTypes", () => {
      it("should return the selected type value array when filters has selected types", () => {
        const result = getters.selectedTypes({
          filters: fixtureFiltersWithSelectedTypes
        });

        expect(result).toMatchSnapshot();
      });

      it("should return an empty array when filters has no selected types", () => {
        const filtersWithNoSelectedTypes = fixtureFilters;

        const result = getters.selectedTypes({
          filters: filtersWithNoSelectedTypes
        });

        expect(result).toMatchObject([]);
      });
    });

    describe("selectedSubTypes", () => {
      it("should return the selected sub-type value array when filters has selected sub-types", () => {
        const filtersWithSubSelectedTypes = fixtureFiltersWithSelectedSubTypes;

        const result = getters.selectedSubTypes({
          filters: filtersWithSubSelectedTypes
        });

        expect(result).toMatchSnapshot();
      });

      it("should return an empty array when filters has no selected sub-types", () => {
        const filtersWithNoSelectedSubTypes = fixtureFilters;

        const result = getters.selectedSubTypes({
          filters: filtersWithNoSelectedSubTypes
        });

        expect(result).toMatchObject([]);
      });
    });

    describe("hasChildren", () => {
      it("should return true for a filter that has children", () => {
        const result = getters.hasChildren({ filters: fixtureFilters })(2);

        expect(result).toBe(true);
      });

      it("should return false for a filter that has no children", () => {
        const result = getters.hasChildren({ filters: fixtureFilters })(3);

        expect(result).toBe(false);
      });
    });
  });

  describe("mutations", () => {
    describe("update", () => {
      it("should update the state filters with the passed value", () => {
        const state = {
          filters: fixtureFilters
        };

        mutations.update(state, fixtureFiltersWithSelectedTypes);

        expect(state.filters).toMatchObject(fixtureFiltersWithSelectedTypes);
      });
    });

    describe("setRootFilter", () => {
      it("should set the state root filter to the value passed", () => {
        const state = {
          rootFilter: null
        };

        mutations.setRootFilter(state, fixtureFilters[1]);

        expect(state.rootFilter).toMatchObject(fixtureFilters[1]);
      });
    });
  });

  describe("actions", () => {
    const context = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      state: {
        filters: fixtureFilters
      }
    };

    afterEach(() => {
      jest.resetAllMocks();
    });

    describe("update", () => {
      it("should commit an update mutation with the passed value", () => {
        actions.update(context, fixtureFiltersWithSelectedTypes);

        expect(context.commit).toHaveBeenCalledTimes(1);
        expect(context.commit).toHaveBeenCalledWith(
          "update",
          fixtureFiltersWithSelectedTypes
        );
      });

      it("should not commit an update mutation when the passed value is identical to the state value", () => {
        actions.update(context, fixtureFilters);

        expect(context.commit).toHaveBeenCalledTimes(0);
      });
    });

    describe("toggle", () => {
      it("should dispatch an update action with the updated filters", () => {
        actions.toggle(context, fixtureFilters[1]);

        expect(context.dispatch).toHaveBeenCalledTimes(1);
        expect(context.dispatch.mock.calls[0][0]).toBe("update");
        expect(context.dispatch.mock.calls[0][1]).toMatchSnapshot();
      });
    });
  });
});
