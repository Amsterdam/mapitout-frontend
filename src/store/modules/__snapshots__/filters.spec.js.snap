// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters store module actions toggle should dispatch an update action with the updated filters 1`] = `
Array [
  Object {
    "iconComponent": "icon-bus",
    "id": 1,
    "name": "Stations",
    "parentId": 0,
    "propertyId": null,
    "selected": false,
    "value": "Station",
  },
  Object {
    "iconComponent": "icon-education",
    "id": 2,
    "name": "Schools",
    "parentId": 0,
    "propertyId": null,
    "selected": true,
    "value": "School",
  },
  Object {
    "group": "School programs",
    "iconComponent": "icon-education",
    "id": 3,
    "name": "Age Range 4-11",
    "parentId": 2,
    "propertyId": 1,
    "selected": false,
    "value": "Age range 4-11",
  },
  Object {
    "group": "School programs",
    "iconComponent": "icon-education",
    "id": 4,
    "name": "Age Range 4-18",
    "parentId": 2,
    "propertyId": 2,
    "selected": false,
    "value": "Age range 4-18",
  },
  Object {
    "group": "School programs",
    "iconComponent": "icon-education",
    "id": 5,
    "name": "Primary Education",
    "parentId": 2,
    "propertyId": 3,
    "selected": false,
    "value": "Primary education",
  },
  Object {
    "group": "School programs",
    "iconComponent": "icon-education",
    "id": 6,
    "name": "Secondary Education",
    "parentId": 2,
    "propertyId": 4,
    "selected": false,
    "value": "Secondary education",
  },
  Object {
    "group": "School system",
    "iconComponent": "icon-education",
    "id": 7,
    "name": "International Schools",
    "parentId": 2,
    "propertyId": 5,
    "selected": false,
    "value": "International schools",
  },
  Object {
    "group": "School system",
    "iconComponent": "icon-education",
    "id": 8,
    "name": "Dutch Schools",
    "parentId": 2,
    "propertyId": 6,
    "selected": false,
    "value": "Dutch schools",
  },
]
`;

exports[`filters store module getters selectedSubTypes should return the selected sub-type value array when filters has selected sub-types 1`] = `
Array [
  "Age range 4-11",
  "International schools",
]
`;

exports[`filters store module getters selectedTypes should return the selected type value array when filters has selected types 1`] = `
Array [
  "School",
]
`;
